3836/3836.md

```diff
@@ -0,0 +1,666 @@
+Its usage is `files.sh (1|2)`, and it prints the location of all calls to error
+in the files comprising the command-line-specified tier. It finds these
+locations via lightweight semantic analysis (read: `grep`).
+
+
+## Tier 1
+
+Tier 1 files contain 17 uses of `error`.
+
+### 1: `ouroboros-network/demo/chain-sync.hs:577`
+```hs
+nextState [] = error "chainSyncServer: impossible"
+```
+
+The list on which `nextState` is called originates from `chainGenerator` in the
+same file. `nextState` will never see an empty list because `chainGenerator`
+creates an infinite list.
+
+As one possible fix, I've replaced the error message with a more informative
+one.
+
+As another, I've lifted the `chainGenerator` result into an `Infinite`
+construction, courtesy of
+[`infinite-list`](https://hackage.haskell.org/package/infinite-list-0.1/docs/Data-List-Infinite.html).
+If this is too much, the commit introducing it (`1189f40e`) can be reverted.
+
+
+### 2: `ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:674`
+```hs
+error $ "A pick policy requested an attribute for peer address "
+     ++ " which is outside of the set given to pick from"
+```
+
+This use of `error` appears in `pickPeers`, which is a "`PickPolicy` wrapper
+function". `pickPeers` is provided:
+- A `PeerSelectionState` containing i.a. `localRootPeers`,
+  `publicRootPeers`, and `knownPeers`
+- A `PickPolicy` called `pick`
+- A `Set peeraddr` called `available`
+
+A `PickPolicy` is "an action that picks a subset of elements from a [collection]
+of peers". It makes this choice by examining each peer in this collection and
+examining them via several functions (generally of the shape `peeraddr -> _`) it
+is provided.
+
+`pickPeers` provides `pick`, its `PickPolicy`, with functions of its own
+creation. Those functions check their input peer for membership in
+`localRootPeers`, `publicRootPeers`, and `knownPeers`, and trigger the error if
+the peer is not found in `knownPeers`, but `pick` will only ever apply them to
+peers in `available`. Therefore, the error may be triggered if a peer in
+`available` is not in `knownPeers` (i.e. if `available` is partially disjoint
+from `knownPeers`).
+
+In practice, `available` appears to be constructed from the various peer
+collections in a `PeerSelectionState`, whose invariant (as described in
+`assertPeerSelectionState` in this same module) ensures that `knownPeers` is a
+superset of all other peer sets, so this error is quite unlikely to be triggered
+as long as any extant `PeerSelectionState` adheres to its invariant.
+
+I've added documentation at this `error` site of the rationale I've derived, and
+have included a reference to the invariant-documenting function at the
+declaration site of `PeerSelectionState`.
+
+I've also added a precondition that checks whether `available` is a subset of
+`knownPeers`. This may offer a poor utility to expense ratio in practice, and
+I'd welcome input and discussion on this change.
```

Comment:
What do you mean by log, exactly? I haven't looked much into the testing facilities for the `PeerSelectionState` API.

Reply:
@coot

Reply:
This would actually be much more involved, it would require to identify all log messages which could  also carry the information about available peers, like `TraceForgetColdPeers`.  Then in a test we can collect all these traces and verify that `available` is a subset of `known` peers.  But this would also pollute the log produced by `cardano-node`, which is not ideal; so we'd need to create another tracer for these values.  This seems to be quite some work, while assertions give us this quite cheaply, so let's do it this way.